design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+```
+
+In this version, the implementations of `createStream(...)` and `createTable(...)` would be identical and could be replaced with a single `createSource(...)`
+but the name of this method feels confusing. As with the previous proposal, `CreateSourceResponse` and `InsertIntoResponse` are identical.
+
+Other alternatives include separating `CREATE <STREAM/TABLE>` and `CREATE <STREAM/TABLE> ... AS SELECT` into separate methods, but that feels unnecessarily complex.
+
+#### `DROP <STREAM/TABLE>`
+
+`Client` methods:
+```
+  /**
+   * Drop stream. The underlying Kafka topic will not be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropStream(String streamName);
+
+  /**
+   * Drop stream. The underlying Kafka topic may optionally be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropStream(String streamName, boolean deleteTopic);
+
+  /**
+   * Drop table. The underlying Kafka topic will not be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropTable(String tableName);
+
+  /**
+   * Drop table. The underlying Kafka topic may optionally be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropTable(String tableName, boolean deleteTopic);
+```
+where `DropSourceResponse` is actually the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse` above.
+
+Again, the implementations of `dropStream(...)` and `dropTable(...)` would be the same so we could instead have a single `dropSource(...)`, but the naming might be confusing.
+
+If we choose to keep them separate, there's an open question of whether the client should validate that `dropStream(...)` is not used to drop a table, and vice versa.
+IMO such validation would be introducing complexity without much benefit, though this point of ambiguity makes me prefer a single `dropSource(...)` if we can agree on a method name that's not confusing.
+
+Note that users can also execute `DROP <STREAM/TABLE>` requests via `executeDdl(...)` or `executDml(...)` above.
+
+#### `TERMINATE <queryId>`
+
+```
+  CompletableFuture<TerminateQueryResponse> terminatePersistentQuery(String queryId);
+```
+where `TerminateQueryResponse` is again the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse`/`DropSourceResponse` above.
+
+The method name `terminatePersistenQuery(...)` is to distinguish from `terminatePushQuery(...)` below.
+
+#### Connectors
+
+```
+  CompletableFuture<ConnectorInfo> createSourceConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<ConnectorInfo> createSinkConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<Void> dropConnector(String name);
+```
+where `ConnectorInfo` is from an Apache Kafka module ([link](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/ConnectorInfo.java)).
+
+Or would we rather not have a dependency on Apache Kafka in the ksqlDB client interfaces?
```

Comment:
+1 I don't think we need connector management as part of the ksqlDB client interface (at least for v1). This seems like something that the connect product should solve and then if the user wants connector management in java they can pull in a dedicated first-class connect client

Reply:
I'll leave this up to @derekjn . My thinking was that if ksqlDB is defined by pull queries + Connect integration, it'd be important for the client to offer both pieces of functionality, though I agree that end state you've described is the more preferable one.

Reply:
Imho, it would be good to have this functionality _at some point_. The more that the user can do with the ksqlDB client without having to use other clients, the better.

However, I'm not convinced we need this _yet_. I'd recommend probably leaving it for now, we can always support it at a later date if we choose to.

Reply:
I agree that this probably isn't something we need right now. Users can always use the REST API directly here as a workaround. Also agree with @purplefox about keeping this on our radar for the future though.

Reply:
Makes sense. Removed connector management from the scope of this KLIP (but left the proposed interfaces at the bottom in case we want to revisit them later).